File: JuegoEnJavayRubyNapakalaki/Napakalaki/src/NapakalakiGame/Napakalaki.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.ArrayList;
import java.util.Random;

/**
 *
 * @author nen155
 */
public class Napakalaki {
    private static final Napakalaki napakalaki =new Napakalaki();
    private Monster currentMonster;
    private CardDealer dealer;
    private ArrayList<Player> players;
    private Player currentPlayer;
    private int indicePlayer;
    private boolean primerJugador;
    

    public static Napakalaki getInstance(){
        return napakalaki;
    }
    

    public CombatResult developCombat(){
        CombatResult r =currentPlayer.combat(currentMonster);
        
        if(r == CombatResult.LoseAndConvert)
        {
            Cultist c = this.dealer.nextCultist();
            CultistPlayer p = new CultistPlayer(this.currentPlayer, c);
            int i = this.players.indexOf(this.currentPlayer);
            this.players.set(i, p);
            this.currentPlayer = p;
        }
        
        
        dealer.giveMonsterBack(currentMonster);
        
        return r;
    }
    public void discardVisibleTreasures(ArrayList<Treasure> treasures){
        for(int i=0;i< treasures.size();i++){
            currentPlayer.discardVisibleTreasure(treasures.get(i));
            dealer.giveTreasureBack(treasures.get(i));
        }
    }
    public void discardHiddenTreasures(ArrayList<Treasure> treasures){
         for(int i=0;i< treasures.size();i++){
            currentPlayer.discardHiddenTreasure(treasures.get(i));
            dealer.giveTreasureBack(treasures.get(i));
        }
    }
    public void makeTreasuresVisible(ArrayList<Treasure> treasures){
        for(int i = 0;i<treasures.size();i++){
            currentPlayer.makeTreasureVisible(treasures.get(i));
        }
    }
    public boolean buyLevels(ArrayList<Treasure> visible,ArrayList<Treasure> hidden){
        return currentPlayer.buyLevels(visible, hidden);
    }
    public void initGame(ArrayList<String> players){
        initPlayers(players);
        dealer.initCards();
        nextTurn();
    }
    public Player getCurrentPlayer(){
        return currentPlayer;
    }
    public Monster getCurrentMonster(){
        return currentMonster;
    }
    public boolean nextTurn(){
        if(nextTurnAllowed()){
            currentMonster = dealer.nextMonster();
            currentPlayer = nextPlayer();
            if(currentPlayer.isDead()){
                currentPlayer.initTreasures();
            }    
            return true;
        }
        else
            return false;
    }
    public boolean endOfGame(CombatResult result){
        if(result.WinAndWinGame == CombatResult.WinAndWinGame){
            return true;
        }else
            return false;
        
    }
    private Napakalaki(){
         //currentMonster= new Monster();
         dealer = CardDealer.getInstance();
         players = new ArrayList();
         primerJugador = true;
        //currentPlayer = nextPlayer();
    }
    private void initPlayers(ArrayList<String> names){
        for(int i=0;i<names.size();i++){
            players.add(new Player(names.get(i)));
        }
        currentPlayer = nextPlayer();
    }
    private Player nextPlayer(){
        Random r = new Random();
        Player p=null;
        if(primerJugador){
            indicePlayer = r.nextInt(players.size()-1);
            primerJugador=false;
        }else{
            
            if(indicePlayer <players.size()-1)
                indicePlayer++;
            else
                indicePlayer=0;
        }
            p = players.get(indicePlayer);
        return p;
    }
    private boolean nextTurnAllowed(){
        return currentPlayer.validState();
    }
}

File: JuegoEnJavayRubyNapakalaki/Napakalaki_Ruby/lib/napakalaki.rb
#encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
#ultima

require_relative 'treasure_kind'
require_relative 'card_dealer'
require_relative 'player'
module Napakalaki
  
  require 'singleton'
  class Napakalaki
    include Singleton
     

    def developCombat
      r=@currentPlayer.combat(@currentMonster)
      
      if (r == CombatResult::LOSEANDCONVERT)
        c = @dealer.nextCultist
        cp = CultistPlayer.new(@currentplayer, c)
        @players.delete(@currentPlayer)
        @players << cp
        @currentPlayer = cp
      end
      
      
        CardDealer.instance.giveMonsterBack(@currentMonster)
      r
    end
    
    def discardVisibleTreasures(treasures)
      if(treasures.size!=0)
      for i in 0..treasures.size-1
            @currentPlayer.discardVisibleTreasure(treasures[i])
            CardDealer.instance.giveTreasureBack(treasures[i])
      end
      end
    end
    def discardHiddenTreasures(treasures)
      if(treasures.size!=0)
      for i in 0..treasures.size-1
            @currentPlayer.discardHiddenTreasure(treasures[i])
            CardDealer.instance.giveTreasureBack(treasures[i])
      end
      end
    end
    def makeTreasuresVisible(treasures)
      if(treasures.size!=0)
      for i in 0..treasures.size-1
            @currentPlayer.makeTreasureVisible(treasures[i]);
      end
      end
    end
    def buyLevels(visible,hidden)
      @currentPlayer.buyLevels(visible,hidden)
    end
    def initGame(players)
      initPlayers(players)
      CardDealer.instance.initCards
      nextTurn
    end
    def getCurrentPlayer
      @currentPlayer
    end
    def getCurrentMonster
      @currentMonster
    end
    def nextTurn
      if (nextTurnAllowed==true)
            @currentMonster = CardDealer.instance.nextMonster
            @currentPlayer = nextPlayer
            if(@currentPlayer.isDead==true)
                @currentPlayer.initTreasures
            end
            true
        else
            false
      end
    end
    def endOfGame(result)
      if(result == CombatResult::WINAANDWINGAME)
           true;
        else
            false;
      end
        
    end
    private
    def initialize
      @primerJugador=true
      @players=Array.new
      @indicePlayer=0
    end
     
     def initPlayers(names)
       @currentPlayer=Player.new(names[0])
       for i in 0..names.size-1
            @players<<Player.new(names[i])
       end
     end
     
     def nextPlayer
        p=nil
        if(@primerJugador==true)
            @indicePlayer = rand(@players.size-1)
            @primerJugador=false
        else
            
            if(@indicePlayer <@players.size-1)
                @indicePlayer=@indicePlayer+1
            else
                @indicePlayer=0
            end
        end
            p = @players[@indicePlayer]
        p
     end
     
     def nextTurnAllowed
       @currentPlayer.validState
     end
    
  end
end
